components/core/src/clp/StringReader.cpp

```diff
@@ -41,6 +41,10 @@ ErrorCode StringReader::try_read(char* buf, size_t num_bytes_to_read, size_t& nu
 }
 
 ErrorCode StringReader::try_seek_from_begin(size_t pos) {
+    if (pos > input_string.size()) {
```

Comment:
I would classify this as a bug I'm fixing in the StringReader class so that I can use it for tests. Every other reader we have will EOF if you seek past the end from what I've seen.

Actually, the way the rest of this class is written if you first seek past the end of the input buffer it will happily let you read data beyond the end of the buffer. I.e. it is very explicitly a bug.

Reply:
Hmm, The FileReader internally calls fseeko, which I believe would allow seeking beyond the end of file.

BufferedFileReader and BufferReader would return ErrorCode_Truncated, and won't update the pos at all if the pos is greater than the maximum length,

From the consistency point of view, maybe we should let it return ErrorCode_Truncated. But also wonder if the rest of your code is dependent on the current behavior.

Reply:
* The implementations of ReaderInterface have inconsistent behaviour.
  * Some (e.g., `FileReader`) will rely on the lower implementation to return an error if the seek fails.
  * Some (e.g. `BufferReader`) will return an error if the seek is past the end, but they won't modify `m_pos`.
  * Some (e.g., `NetworkReader`) will return an error if the seek is past the end, and they will modify `m_pos`.
  * Each of the above returns different error codes.

I think the practical implementation is for `try_seek_from_begin` to:
* try to seek until `pos`
* if that's past the end of the medium (file/buffer/etc.), `m_pos` should be updated to just past the last byte.
* the method should return `ErrorCode_EndOfFile`.

The reason to update the `pos` even though we don't get 

If y'all agree, we should open a GH issue to refactor the existing implementations. And for this implementation, we implement the proposal above (basically what Devin's implemented---although on error, we're still updating `m_cur_pos`, even tough the error may not be EOF?).

Reply:
Sure, the standard behavior sounds good to me.

Reply:
That makes sense to me.

Also went and updated BoundedReader to only update `m_curr_pos` on error if that error is EOF.